tiled/client/sync.py

```diff
@@ -0,0 +1,112 @@
+import collections.abc
+import itertools
+
+from ..structures.core import StructureFamily
+from ..structures.data_source import DataSource, Management
+from .base import BaseClient
+
+
+def sync(
+    source: BaseClient,
+    dest: BaseClient,
+):
+    """
+    Copy data from one Tiled instance to another.
+
+    Parameters
+    ----------
+    source : tiled node
+    dest : tiled node
+
+    Examples
+    --------
+
+    Connect to two instances and copy data.
+
+    >>> from tiled.client import from_uri
+    >>> a = from_uri("http://localhost:8000", api_key="secret")
+    >>> b = from_uri("http://localhost:9000", api_key="secret")
+    >>> sync(a, b)
+
+
+    Copy select data.
+
+    >>> sync(a.items().head(), b)
+    >>> sync(a.search(...), b)
+
+    """
+    if hasattr(source, "structure_family"):
+        # looks like a client object
+        _DISPATCH[source.structure_family](source.include_data_sources(), dest)
+    elif isinstance(source, list):
+        # such as result of source.items().head()
+        _DISPATCH[StructureFamily.container](dict(source), dest)
+    elif isinstance(source, collections.abc.Mapping):
+        _DISPATCH[StructureFamily.container](source, dest)
+
+
+def _sync_array(source, dest):
+    num_blocks = (range(len(n)) for n in source.chunks)
+    # Loop over each block index --- e.g. (0, 0), (0, 1), (0, 2) ....
+    for block in itertools.product(*num_blocks):
+        array = source.read_block(block)
+        dest.write_block(array, block)
+
+
+def _sync_table(source, dest):
+    for partition in range(source.structure().npartitions):
+        df = source.read_partition(partition)
+        dest.write_partition(df, partition)
+
+
+def _sync_container(source, dest):
```

Comment:
This function is more complex than I would like but after about 20 minutes of fiddling I couldn't land of a significantly simpler implementation. Might  be worth revisiting later.